### PR TITLE
[doc][build] Stop using Bazel and change commit trace logic

### PR DESCRIPTION
**Pull request by khluu**
- Change the logic to find commit to fetch cache from. Previously, it was trying to find common ancestor of current commit and `origin/master` but this would fail if users use ray fork. Changed the logic to go through commits in the log and find the most recent one that already has the cache on S3.
- Stop using bazel and move the scripts to `doc/` directory
- Add extra dependencies to `requirements-docs` since these were previously attached with Bazel run.

### CODE REVIEW

File: doc/Makefile [hunk 1]
```diff
@@ -82,8 +82,8 @@ develop: html
 RAY_DIR := $(shell pwd | rev | cut -d'/' -f2- | rev)
 
 local:
-	bazel run //ci/ray_ci/doc:cmd_load_doc_cache -- --ray-dir=$(RAY_DIR)
-	bazel run //ci/ray_ci/doc:cmd_update_cache_env -- --ray-dir=$(RAY_DIR)
```
Comment (aslonnie) [inline doc/Makefile:86]: this should have been using bazelisk

or use some prebuilt python executable binary / zip

**Review by aslonnie — COMMENTED** [review]
why stop using bazel? people will still need to install the required python dependencies right? so why not just ask people to install bazel instead?

**Comment by khluu** [comment]
> why stop using bazel? people will still need to install the required python dependencies right? so why not just ask people to install bazel instead?

Even if they already have Bazel installed, we've seen a few cases where it took Bazel a while to start up & load all its dependencies before fetching doc cache, which kind of defeats the purpose of using cache is to make build faster. If they install the packages in their conda environment, they don't have to wait for it.

**Comment by aslonnie** [comment]
> If they install the packages in their conda environment, they don't have to wait for it.

it also takes time to install the conda environment, right? which also defeats the purpose?

it's ultimately up to you, I am just saying:

- you need a stable way to install and maintain the execution environment and the dependency required anyways. I am not sure why conda is within the assumption but bazel is not. this seems to be creating a different dev environment that is different from ray's standard one.
- if you really care about the boot up latency, you should maybe consider ship a binary with no dependencies from the environment, either with a python executable zip, or a go binary (e.g. via the [rayci](https://github.com/ray-project/rayci) repo's releases).

**Comment by aslonnie** [comment]
or, you can also remove the 3rd-party dependencies from the python file (remove click and botocore), and just use python std libs, it still depends on the python interpreter, but that is a reasonable thing to assume that a user should have.

and then probably should just move the python files to the doc directory, and do not keep them in the ci directory. so that it does not need `../` find it.

**Comment by can-anyscale** [comment]
I think the main complains we have from users is that we are adding another dependencies to their existing workflow (not saying their existing workflow of using conda and requirement is perfect), and they hesitate to change.

I also think we should either use conda or bazel and not both (they duplicate the jobs of each other so creating wasted overhead), but decide to stick with conda since it's the existing users workflow.

**Comment by aslonnie** [comment]
bazel should have been already in their workflows.. but anyways..

maybe just:

- move the commands out of the `ci` directory. if it is not going to be used as a bazel `py_binary`, then don't pretend that it is.
- make it clear how to make dependency changes to the python script, and test it somewhere on CI that with `pip install doc/requirements-doc.txt` and the script can run without issue.

**Comment by can-anyscale** [comment]
You're right that internal engineers have bazel but when do testing with folks like technical writer and external contributors, since they don't have to build ray, they don't have bazel in their environments. Some also has different bazel version installed (for a different project I assume), so ray refused to use it and we need to ask them to install bazelisk. 

This small problem leads to the sentiment is that this new tool is unreliable (w00t).

So this PR is the compromise between minimizing the changes to an existing users' workflow, and the time we need to invest in this project to have a good enough solution (if we have more time, I think convert the whole thing to bazel and re-educate users could be worthwhile).

**Review by aslonnie — APPROVED** [review]

**Review by can-anyscale — APPROVED** [review]

File: ci/ray_ci/doc/BUILD.bazel [hunk 2]
```diff
@@ -115,19 +93,3 @@ py_test(
         ci_require("pytest"),
     ],
 )
-
-py_test(
-    name = "test_update_cache_env",
```
Comment (can-anyscale) [inline ci/ray_ci/doc/BUILD.bazel:120]: we lost this test now

**Pull request merged**